## Rubric

Instructions: DELETE this cell before you submit via a `git push` to your repo before deadline. This cell is for your reference only and is not needed in your report. 

 Scoring: Out of 10 points

- Each Developing  => -2 pts
- Each Unsatisfactory/Missing => -4 pts
  - until the score is 0

If students address the detailed feedback in a future checkpoint they will earn these points back






|                                  | **Unsatisfactory**                                                                                                                                                                                                                                                                                                                        | **Developing**                                                                                                                                                                                                       | **Proficient**                                                                                                                                                                                            | **Excellent**                                                                                                                                                                            |
|----------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **EDA relevance**                | EDA is mostly neither relevant to the question nor helpful in figuring out how to address the question. Or the EDA does address the question, but many obviously relevant variables / analyses / figures were not included. | EDA is partly irrelevant/unhelpful. EDA missed one or two obvioulsy relevant analysis (distributions of single variables or relationships between variables) | EDA includes the obviously relevant / helpful variables in addressing the question.                                                              | Thorough EDA fully explored the dataset                                                                                                                 |
| **EDA analysis and description** | Many of the analyses are poor choices (e.g., using means instead of medians for obviously skewed data), or are poorly described in the text, or do not aid understanding the data                                                                                                                                                     | Some of the analyses are poor choices, or are poorly described in the text, or do not aid understanding the data                                                                                                 | All analyses are correct choices. Only one or two have minor issues in the text descriptions supporting them. Mostly they fit well with other elements of the EDA and support understanding the data  | All analyses are correct choices with clear text descriptions supporting them. The figures fit well with the other elements of the EDA, producing a clear understanding of the data. |
| **EDA figures**                  | Many of the figures are poor plot choices (e.g., using a bar plot to represent a time series where it would be better to use a line plot) or have poor aesthetics (including colormap, data point shape/color, axis labels, titles, annotations, text legibility) or do not aid understanding the data                                | Some of the figures are poor plot choices or have poor aesthetics. Some figures do not aid understanding the data                                                                                                | All figures are correct plot choices. Only one or two have minor questionable aesthetic choices. The figures mostly fit well with the other elements of the EDA and support understanding the data    | All figures are correct plot choices with beautiful aesthetics. The figures fit well with the other elements of the EDA, producing a clear understanding of the data.                |





# COGS 108 - EDA Checkpoint

## Authors

Instructions: REPLACE the contents of this cell with your team list and their contributions. Note that this will change over the course of the checkpoints

This is a modified [CRediT taxonomy of contributions](https://credit.niso.org). For each group member please list how they contributed to this project using these terms:
> Analysis, Background research, Conceptualization, Data curation, Experimental investigation, Methodology, Project administration, Software, Visualization, Writing – original draft, Writing – review & editing

Example team list and credits:
- Alice Anderson: Conceptualization, Data curation, Methodology, Writing - original draft
- Bob Barker:  Analysis, Software, Visualization
- Charlie Chang: Project administration, Software, Writing - review & editing
- Dani Delgado: Analysis, Background research, Visualization, Writing - original draft

# Research Question

To what extent has the variability in mainstream music, measured by the difference in features like tempo, key, valence, explicitness, energy, loudness of the monthly top three Billboard Hot 100 songs, increased or decreased between the streaming era (2013-2026) relative to the pre-streaming era (2000-2012)?

(We chose this time epoch because 2013 was the beginning of the streaming-dominant era, as it generally marks the first year in which streaming became the primary mode of music consumption in the United States. To allow for a balanced comparison, we chose to analyze an equal span of time before (2000-2012) and after (2013-2026) this shift.)


## Background and Prior Work

As digitalization has rapidly advanced, no stone has been left unturned. The way we listen to music has been one of the biggest changes; progressing from physical cassettes, records, and CDs to singular devices that hold hundreds of millions of tracks. This unparalleled access to music has not come lightly. Beginning in the 2010s, music consumers are able to purchase or subscribe to a streaming platform to gain access to a dataset of songs, readily available to them. Since the release of streaming platforms like Spotify and Apple Music, the barrier of entry to consume and release music has lowered exponentially.

However, the way we consume music has also changed, with the algorithm becoming the ruling factor for many streaming platforms and social media sites. The more a song appeals to an algorithm, the more streams a song has, and all the more money it makes. This cause and effect has led to what we believe has been a homogenization of music since the streaming age began, where songs are designed for mass appeal. In this project, we’d like to explore whether or not music has become more homogenized when comparing post-streaming (2013-2026) and pre-streaming (2000-2013) eras. Our definition of homogeneity will include four features: energy, valence, acousticness, instrumentalness. Energy measures the "danceability" and animation of a given song, faster tempos and louder volumes connote higher scales of danceability and activity in a track, whereas softer, quieter volumes and slower tempos connote lower scales of energy. Valence describes the musical positivity, higher scales of valence include upbeat, happier, and euphoric music while lower values will include more moody, sadder, and tense musical qualities. High acousticness will include songs that sound organically produced physically by humans with instruments and lower values of acousticness will include songs that sound electronically produced, including techno-music. For example, more electronic music highlights faster tempos, which will be measured as in beats per minute (BPM). Instrumentalness highlights how many words are in a given track. A song with many lyrics will be graded lower on the instrumentalness scale, whereas tracks with mostly instrumental portions as very few lyrics and vocals will be graded higher on the scale. These four features: energy, valence, acousticness, and instrumentalness, are indicators of relative homogeneity. In our project, we would like to see if these features converge or diverge among the monthly top three Billboard Hot 100 songs during the streaming era (2013-2026) relative to the pre-streaming era (200-2012).

Perhaps one of the clearest signs of how music has changed is the change in the average length of music over time, which has decreased by more than 30 seconds over time. Many of the top songs of the past few years have clocked in at under three minutes, a clear departure from longer music. This is most likely tied to the difference streaming makes when considering song structure. Researchers have found a correlation between skipping music and song structure where users are likely to skip music before the song has even made it to 30 seconds. Since streaming services only pay artists for “plays” and “plays” only count if a song has been played for longer than 30 seconds, it stands to reason that many artists and songwriters have focused their energy towards creating catchy music that’s interesting enough to capture attention, but short enough that listeners won’t get bored and skipping, leaving artists without their hard-earned revenue.

In a previous study from Tilburg University, the impact of streaming services on homogenized mainstream music is mixed. In their research, Knox and Datta look at the supply and demand side of the industry, focusing on the “acoustic diversity” of a given song: the pitch, loudness, valence, tempo, and duration. In the early beginnings of platforms like Napster, an illegal music sharing platform, YouTube, and Spotify, artists were able to easily create and distribute their music to a large audience in a way that had not been seen before. Musicians tended to mimic each other in the early stages of the new widespread distribution of music, leading to an initial decrease in diversity in music quality. However, when streaming services like Spotify became more popularized, algorithms in place were able to provide new recommendation systems and search tools that enabled listeners to discover a vast amount of music readily available to them. From a demand standpoint, consumers have more choices of music selection available to them, thus allowing for more diversity within their consumption.

Despite the initial dip in acoustic diversity, there has been a weak, steady increase in acoustic diversity since the release of streaming platforms due to the large catalog of recorded music available. However, our project is looking to see if special media platforms like TikTok or Instagram influence the homogeneity of music selections within streaming platforms. We aim to explore this topic by focusing on the categorization of songs through factors such as tempo, lyrics, etc, that were on the Billboard Top 100 over the last 60-70 years. Through our exploration and analysis of this data, we hope to discover whether streaming has truly homogenized popular music or if streaming has had little to no effect on the variability of popular music.

^ Rosenblatt, Bill. “Streaming Is Changing the Sound of Music.” The Wall Street Journal, 15 Sept. 2023, www.wsj.com/arts-culture/music/streaming-is-changing-the-sound-of-music-182dc907?gaa_at=eafs&gaa_n=AWEtsqfipow1GoFn2kJm7IBA7Gz4TYoUnk-sSGEj-DU5EpPTInPq82ZxQwundm4CAOg%3D&gaa_ts=6983c67b&gaa_sig=l3vMKS7DG8OgIs-7w4HPTc6jkRjV7wPNkNWGRtGIdoiv5nR9ZQakPh5m8NDaYWNwG5YKd7nUowkS270C-9jYhg%3D%3D. Accessed 4 Feb. 2026. ^ Montecchio, Nicola, et al. “The Skipping Behavior of Users of Music Streaming Services and Its Relation to Musical Structure.” PLOS ONE, edited by Jichang Zhao, vol. 15, no. 9, Sept. 2020, p. e0239418, https://doi.org/10.1371/journal.pone.0239418. ^ Knox, G., & Datta, H. (2020). Streaming Services and the Homogenization of Music Consumption.

# Hypothesis


We hypothesize that songs from the years 2013-2026, which would be considered the streaming era, will exhibit less variability in measurable factors such as danceability, energy valence, acousticness, instrumentalness, tempo, duration, loudness, key/mode, explicitness, and speechness when comparing them to songs from 2000-2012, which would be the pre-streaming era. This would indicate that popular songs over the years have become more homogenized. This could be for a wide variety of reasons, but we theorize that it is largely because of two reasons. The first is because of corporate involvement with music creation. Music corporations want to maximize profits and popularity of an artist's music, which makes them push for their artists to create music that is broader, more generic, and more formulaic, which has led popular songs to sound more similar in recent years. This is compounded by the second cause, the rise of streaming services. Streaming services are operated as a business that pits artists against each other in a popularity contest. More streams mean an artist's music is pushed out more, which means more money for the artist and exponentially more for the business. This also encourages streaming services to promote music that appeals to the broadest audience, as it generates the most revenue for them.

## Data

### Data overview

- Dataset #1:

    - Dataset Name: 160k Spotify songs from 1921 to 2020 (Sorted)
    - Link to the dataset: https://github.com/fcpercival/spotify-data-sorted
    - Number of observations: 160,000+
    - Number of variables: 12
    - Description of the variables most relevant to this project: Danceability (unitless, 0–1) estimates how suitable a track is for dancing. Energy (0–1) measures intensity and activity. Valence (0–1) describes musical positivity. Acousticness (0–1) estimates the likelihood a track is acoustic rather than electronically produced. Instrumentalness (0–1) predicts whether a track contains vocals Tempo is measured in beats per minute (BPM) Duration is given in milliseconds; dividing by 1000 converts to seconds. Loudness is measured in decibels (dB) Key is encoded as an integer 0–11 representing the 12 pitch classes (C through B) Mode indicates major (1) or minor (0). Speechiness (0–1) measures the presence of spoken words
    - Descriptions of any shortcomings this dataset has with repsect to the project: Dataset doesnt have data past 2020 and can reflect biases from Spotify's reccomendation algorithms.

  
- Dataset #2

    - Dataset Name: Hit Songs Prediction Dataset Link to the dataset: https://github.com/Orfium/hsp-lyrics-dataset?tab=readme-ov-file
    - Number of observations: ~12,000
    - Number of variables: 8 relevant
    - Description of the variables most relevant to this project: Danceable, alternative, electronic, pop, aggressive, happy, bright, and instrumental. All measured on a scale from 0-1. Indications of genre and mood surrounding the song.
    - Descriptions of any shortcomings this dataset has with repsect to the project: Does not include any data from past 2010, which does not include the post streaming era we intend to target.

  
- Dataset #3

    - Dataset Name: BillBoard Hot100
    - Link to the dataset: https://www.kaggle.com/datasets/ludmin/billboard?select=hot100.csv
    - Number of observations: ~350,000
    - Number of variables: 7
    - Description of the variables most relevant to this project: chart_week = date/week of the chart. this_week = rank that week. title = song name. performer = artist name. last_week = rank previous week. peak_pos = highest rank reached. wks_on_chart = weeks on the Hot 100 chart.
    - Descriptions of any shortcomings this dataset has with repsect to the project: Only includes Billboard's rankings, which are mainstream media centralized to the United States and North America.

In [1]:
# Run this code every time when you're actively developing modules in .py files.  It's not needed if you aren't making modules
#
## this code is necessary for making sure that any modules we load are updated here 
## when their source code .py files are modified

%load_ext autoreload
%autoreload 2

In [24]:
# Setup code -- Run only once after cloning!!! 
#
# this code downloads the data from its source to the `data/00-raw/` directory
# if the data hasn't updated you don't need to do this again!

# if you don't already have these packages (you should!) uncomment this line
# %pip install requests tqdm

import sys
sys.path.append('./modules') # this tells python where to look for modules to import

import get_data # this is where we get the function we need to download data

# replace the urls and filenames in this list with your actual datafiles
# yes you can use Google drive share links or whatever
# format is a list of dictionaries; 
# each dict has keys of 
#   'url' where the resource is located
#   'filename' for the local filename where it will be stored 
datafiles = [
    { 'url': 'https://raw.githubusercontent.com/FCPercival/spotify-data-sorted/refs/heads/master/data.csv', 'filename':'music-data.csv'},
]

get_data.get_raw(datafiles,destination_directory='data/00-raw/')

Overall Download Progress: 100%|██████████| 1/1 [00:00<00:00,  1.29it/s]

Successfully downloaded: music-data.csv


### Dataset #1 :BillBoard Hot100: Top 50 Annually

Scraped from Wikipedia: https://en.wikipedia.org/wiki/Billboard_Year-End_Hot_100_singles_of_1958

This dataset comprises of Billboard's annual overall ranking of the top 50 songs since 1958, including song title, artist, and rank (from 1-50). This dataset is the most up-to-date one of BillBoard chart data, but instead of going up to January/February 2026, its last recorded data ends at December of 2025. For the purposes of this data analysis, however, we will generalize the conclusions from song post 2014 to October 2025 to be part of our "post-streaming (2013-2026)" dataset, due to the large size of the dataset compared to the relatively small time period of missing values. 

The dataset has been filtered below for the appropriate years (2000-2025), and divided into separate dataframes sorted based on year and chart rank, taking the top 10 out the top 50 as our sample space. We chose to analyze starting from 2000 instead of from the inception of the Billboard Top 100 chart in 1958 because it would normalize the number of data points we have from our pre-streaming and post-streaming eras. 

In [25]:
# run pip install lxml if not already

In [43]:
import requests
import pandas as pd
import time
from io import StringIO

start_year = 2000
end_year = 2025

all_years = []

headers = {
    "User-Agent": "Mozilla/5.0"
}

for year in range(start_year, end_year + 1):
    try:
        url = f"https://en.wikipedia.org/wiki/Billboard_Year-End_Hot_100_singles_of_{year}"
        
        response = requests.get(url, headers=headers)
        response.raise_for_status()

        html = StringIO(response.text)
        tables = pd.read_html(html)

        for table in tables:
            if "Title" in table.columns or "Song" in table.columns:
                df = table
                break

        df = df.head(10).copy()

        df = df.iloc[:, -2:]
        
        df.columns = ["song", "artist"]
        
        df["year"] = year
        
        df = df[["year", "song", "artist"]]
        
        df.columns = [str(col).lower() for col in df.columns]

        if "title" in df.columns:
            df = df.rename(columns={"title": "song"})
        if "artist(s)" in df.columns:
            df = df.rename(columns={"artist(s)": "artist"})

        if "song" not in df.columns:
            df.rename(columns={df.columns[1]: "song"}, inplace=True)
        if "artist" not in df.columns:
            df.rename(columns={df.columns[2]: "artist"}, inplace=True)

        all_years.append(df)

        print(f"Finished {year}")

    except Exception as e:
        print(f"Skipped {year}: {e}")

final_df = pd.concat(all_years, ignore_index=True)

print(final_df.head())

Finished 2000
Finished 2001
Finished 2002
Finished 2003
Finished 2004
Finished 2005
Finished 2006
Finished 2007
Finished 2008
Finished 2009
Finished 2010
Finished 2011
Finished 2012
Finished 2013
Finished 2014
Finished 2015
Finished 2016
Finished 2017
Finished 2018
Finished 2019
Finished 2020
Finished 2021
Finished 2022
Finished 2023
Finished 2024
Finished 2025
   year                   song                             artist
0  2000              "Breathe"                         Faith Hill
1  2000               "Smooth"       Santana featuring Rob Thomas
2  2000          "Maria Maria"  Santana featuring The Product G&B
3  2000         "I Wanna Know"                                Joe
4  2000  "Everything You Want"                   Vertical Horizon


In [44]:
final_df.to_csv('data/01-interim/billboard_top10_annual')

top10_yearly = final_df

def determine_era(year):
    if year < 2013:
        return "pre"
    else:
        return "post"

top10_yearly["era"] = top10_yearly["year"].apply(determine_era)

display(top10_yearly.isna().sum())

year      0
song      0
artist    0
era       0
dtype: int64

In [45]:
top10_yearly["song"] = top10_yearly["song"].astype("string")
top10_yearly["artist"] = top10_yearly["artist"].astype("string")
print(top10_yearly.dtypes)

year               int64
song      string[python]
artist    string[python]
era               object
dtype: object


In [46]:
#artist cleanup (will be applied to Spotify 160k later as well)
import re

def normalize_artists(artist_string):
    if pd.isna(artist_string):
        return artist_string
    artist_string = artist_string.lower()
    artist_string = re.sub(r'featuring|feat\.?|with|&| x | and ', ',', artist_string) 
    artist_string = re.sub(r'\s*,\s*', ',', artist_string)
    artist_string = re.sub(r',+', ',', artist_string)
    artist_list = artist_string.split(',')
    artist_list = [a.strip() for a in artist_list if a.strip()]
        
    return ",".join(artist_list)


In [47]:
#apply artist cleaning
top10_yearly["artists"] = top10_yearly["artist"].apply(normalize_artists)
top10_yearly.drop(columns=['artist'], inplace=True)

top10_yearly

,year,song,era,artists
0,2000,"""Breathe""",pre,faith hill
1,2000,"""Smooth""",pre,"santana,rob thomas"
2,2000,"""Maria Maria""",pre,"santana,the product g,b"
3,2000,"""I Wanna Know""",pre,joe
4,2000,"""Everything You Want""",pre,vertical horizon
...,...,...,...,...
255,2025,"""Beautiful Things""",post,benson boone
256,2025,"""Ordinary""",post,alex warren
257,2025,"""I Had Some Help""",post,"post malone,morgan wallen"
258,2025,"""APT.""",post,"rosé,bruno mars"


In [48]:
top10_2000_2013 = top10_yearly[top10_yearly["era"] == "pre"]
top10_2000_2013.shape

(130, 4)

In [49]:
top10_2014_2026 = top10_yearly[top10_yearly["era"] == "post"]
top10_2014_2026.shape

(130, 4)

In [50]:
top10_2000_2012.to_csv('data/02-processed/pre_streaming_top10_annual')
top10_2013_2025.to_csv('data/02-processed/post_streaming_top10_annual')

### Dataset #2: 160K Spotify Songs (1921-2020)

https://github.com/fcpercival/spotify-data-sorted

The 160K Spotify Songs (1921–2020) is a large music metadata collection derived from Spotify’s Web API. It contains roughly 160,000 tracks and includes artist information, genre labels, release year, and several audio features.

The most important metrics are Spotify’s audio features, which are standardized numeric descriptors. Danceability (unitless, 0–1) estimates how suitable a track is for dancing. Values closer to 1 indicate more danceable songs. Energy (0–1) measures intensity and activity; higher values correspond to louder and faster tracks. Valence (0–1) describes musical positivity. Higher values sound happier or more euphoric, while lower values tend to sound sadder or more tense. Acousticness (0–1) estimates the likelihood a track is acoustic rather than electronically produced. Instrumentalness (0–1) predicts whether a track contains vocals; values above ~0.5 suggest mostly instrumental music. These features are probabilities or model outputs rather than physical measurements, so they are relative indicators rather than exact physical quantities.

Several features are expressed in real physical units. Tempo is measured in beats per minute (BPM); most contemporary pop songs fall roughly between 90–130 BPM, while extreme values (e.g., >200 BPM) indicate very fast genres like speed metal or certain electronic styles. Duration is given in milliseconds; dividing by 1000 converts to seconds. Loudness is measured in decibels (dB), typically ranging from about −60 dB (very quiet) to 0 dB (maximum digital level). Key is encoded as an integer 0–11 representing the 12 pitch classes (C through B), and mode indicates major (1) or minor (0). Speechiness (0–1) measures the presence of spoken words; values above ~0.66 are likely spoken word (e.g., podcasts, rap verses heavy in speech), while values below ~0.33 are more purely musical. These metrics allow analysis of musical style, mood, production trends, and historical changes over time.

The primary concern with bias in this dataset is derived from the fact that the data is taken from Spotify’s catalog and recommendation algorithms, meaning it reflects Spotify’s licensing agreements and popularity trends, not a perfectly representative sample of all music ever recorded. Earlier decades (e.g., 1920s–1950s) are likely underrepresented and skewed toward recordings that have been digitized and re-released, which may also show biases. Furthermore, the data does not consist of music from 2020-2026, which is a large segment of the time period we are analyzing for post-streaming.
 as above, add any more copies of this that you need to given how many datasets you have

In [51]:
import pandas as pd
import os
from sklearn.preprocessing import StandardScaler, MinMaxScaler

df = pd.read_csv("data/00-raw/music-data.csv")
drops = ['liveness', 'popularity']
df = df.drop(columns=drops)
print(df.columns)
print(df.isnull().sum())

df["duration_min"] = df["duration_ms"] / 60000
df = df.drop(columns=["duration_ms"])

features = [
    "acousticness",
    "danceability",
    "energy",
    "instrumentalness",
    "speechiness",
    "tempo",
    "valence",
    "duration_min"
]

X = df[features].copy()
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled_df = pd.DataFrame(X_scaled, columns=features)

loudness_scaled = MinMaxScaler().fit_transform(df[["loudness"]])

df_scaled = X_scaled_df.copy()
df_scaled["loudness"] = loudness_scaled
df[df_scaled.columns] = df_scaled

folder_name = 'data/01-interim'
file_name = 'filtered_feature_map.csv'
full_path = os.path.join(folder_name, file_name)

if not os.path.exists(folder_name):
    os.makedirs(folder_name, exist_ok=True)

df.to_csv(full_path, index=False)

Index(['id', 'name', 'artists', 'duration_ms', 'release_date', 'year',
       'acousticness', 'danceability', 'energy', 'instrumentalness',
       'loudness', 'speechiness', 'tempo', 'valence', 'mode', 'key',
       'explicit'],
      dtype='object')
id                  0
name                0
artists             0
duration_ms         0
release_date        0
year                0
acousticness        0
danceability        0
energy              0
instrumentalness    0
loudness            0
speechiness         0
tempo               0
valence             0
mode                0
key                 0
explicit            0
dtype: int64


In [52]:
df

,id,name,artists,release_date,year,acousticness,danceability,energy,instrumentalness,loudness,speechiness,tempo,valence,mode,key,explicit,duration_min
0,0gNNToCW3qjabgTyBSjt3H,!Que Vida! - Mono Version,['Love'],11/1/66,1966,0.084389,0.352754,0.192262,-0.513657,0.754788,-0.408559,0.291280,0.056816,1,9,0,-0.089407
1,0tMgFpOrXZR6irEOLNWwJL,"""40""",['U2'],2/28/83,1983,-0.704189,-0.970359,-0.032130,1.762074,0.757889,-0.423232,1.081074,-0.739660,1,8,0,-0.606379
2,2ZywW3VyVx6rrlrX75n3JB,"""40"" - Live",['U2'],8/20/83,1983,-1.044579,-1.517854,0.730801,-0.476641,0.787284,-0.290510,0.850430,-0.964503,1,8,0,-0.042920
3,6DdWA7D1o5TU2kXWyCLcch,"""40"" - Remastered 2008",['U2'],2/28/83,1983,-0.818360,-0.953250,0.210961,1.357973,0.793751,-0.422565,1.085956,-0.846365,1,8,0,-0.607805
4,3vMmwsAiLDCfyc1jl76lQE,"""40"" - Remastered 2008",['U2'],2/28/83,1983,-0.818360,-0.953250,0.210961,1.357973,0.793751,-0.422565,1.085956,-0.846365,1,8,0,-0.607805
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
169902,4KiYPYpm4ultIp247ftPlI,텅 빈 마음 Empty Heart,['LEE SEUNG HWAN'],10/15/89,1989,-0.170505,0.050491,-0.690344,-0.523517,0.715277,-0.403223,-1.392387,-1.429431,0,11,0,0.152371
169903,42tFTth2jcF7iSo0RBjfJF,피카부 Peek-A-Boo,['Red Velvet'],11/17/17,2017,-1.079096,1.715788,1.546089,-0.515208,0.883063,-0.269835,-0.064922,0.407418,0,1,0,-0.349130
169904,6XP9L7di5JnOc9WaeAW8oe,행복 (Happiness),['Red Velvet'],8/1/14,2014,-0.451950,1.077044,1.430153,-0.523497,0.891097,0.426455,0.146130,0.567475,0,10,0,-0.085888
169905,3Gpdzw72aBVJSrm5J1leVK,"헤어지지 못하는 여자, 떠나가지 못하는 남자 Can't Breakup Girl, C...","['Leessang', 'Jung In']",10/6/09,2009,-0.911291,-0.086383,1.317958,-0.523517,0.900399,1.580268,1.853587,-0.568170,0,10,0,0.438667


The following section cleans up the Spotify 160k dataset to only include those in out Billboard Top 50 Annual sample set

In [93]:
import pandas as pd
import re
import ast

def clean_title(title):
    if pd.isna(title):
        return title
    
    title = title.lower()
    
    # remove quotes
    title = title.replace('"', '')
    title = title.replace("'", "")
    
    # remove parentheses content
    title = re.sub(r"\(.*?\)", "", title)
    
    # remove dash versions
    title = re.sub(r"-.*", "", title)
    
    # remove punctuation
    title = re.sub(r"[^\w\s]", "", title)
    
    return title.strip()

def clean_spotify_artists(artist_string):
    if pd.isna(artist_string):
        return None
    
    # convert string list → python list
    artist_list = ast.literal_eval(artist_string)
    
    # lowercase and strip
    artist_list = [a.lower().strip() for a in artist_list]
    
    return ",".join(artist_list)

In [94]:
pre_billboard = pd.read_csv("data/02-processed/pre_streaming_top10_annual")
post_billboard = pd.read_csv("data/02-processed/post_streaming_top10_annual")

pre_billboard["song"].dtype

dtype('O')

In [95]:
pre_billboard["song_clean"] = pre_billboard["song"].apply(clean_title)
post_billboard["song_clean"] = post_billboard["song"].apply(clean_title)
df["song_clean"] = df["name"].apply(clean_title)

df["artist_clean"] = df["artists"].apply(normalize_artists)
df["artist_clean"] = df["artists"].apply(clean_spotify_artists)

In [111]:
spotify_df = df
spotify_df

,id,name,artists,release_date,year,acousticness,danceability,energy,instrumentalness,loudness,speechiness,tempo,valence,mode,key,explicit,duration_min,song_clean,artist_clean
0,0gNNToCW3qjabgTyBSjt3H,!Que Vida! - Mono Version,['Love'],11/1/66,1966,0.084389,0.352754,0.192262,-0.513657,0.754788,-0.408559,0.291280,0.056816,1,9,0,-0.089407,que vida,love
1,0tMgFpOrXZR6irEOLNWwJL,"""40""",['U2'],2/28/83,1983,-0.704189,-0.970359,-0.032130,1.762074,0.757889,-0.423232,1.081074,-0.739660,1,8,0,-0.606379,40,u2
2,2ZywW3VyVx6rrlrX75n3JB,"""40"" - Live",['U2'],8/20/83,1983,-1.044579,-1.517854,0.730801,-0.476641,0.787284,-0.290510,0.850430,-0.964503,1,8,0,-0.042920,40,u2
3,6DdWA7D1o5TU2kXWyCLcch,"""40"" - Remastered 2008",['U2'],2/28/83,1983,-0.818360,-0.953250,0.210961,1.357973,0.793751,-0.422565,1.085956,-0.846365,1,8,0,-0.607805,40,u2
4,3vMmwsAiLDCfyc1jl76lQE,"""40"" - Remastered 2008",['U2'],2/28/83,1983,-0.818360,-0.953250,0.210961,1.357973,0.793751,-0.422565,1.085956,-0.846365,1,8,0,-0.607805,40,u2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
169902,4KiYPYpm4ultIp247ftPlI,텅 빈 마음 Empty Heart,['LEE SEUNG HWAN'],10/15/89,1989,-0.170505,0.050491,-0.690344,-0.523517,0.715277,-0.403223,-1.392387,-1.429431,0,11,0,0.152371,텅 빈 마음 empty heart,lee seung hwan
169903,42tFTth2jcF7iSo0RBjfJF,피카부 Peek-A-Boo,['Red Velvet'],11/17/17,2017,-1.079096,1.715788,1.546089,-0.515208,0.883063,-0.269835,-0.064922,0.407418,0,1,0,-0.349130,피카부 peek,red velvet
169904,6XP9L7di5JnOc9WaeAW8oe,행복 (Happiness),['Red Velvet'],8/1/14,2014,-0.451950,1.077044,1.430153,-0.523497,0.891097,0.426455,0.146130,0.567475,0,10,0,-0.085888,행복,red velvet
169905,3Gpdzw72aBVJSrm5J1leVK,"헤어지지 못하는 여자, 떠나가지 못하는 남자 Can't Breakup Girl, C...","['Leessang', 'Jung In']",10/6/09,2009,-0.911291,-0.086383,1.317958,-0.523517,0.900399,1.580268,1.853587,-0.568170,0,10,0,0.438667,헤어지지 못하는 여자 떠나가지 못하는 남자 cant breakup girl cant...,"leessang,jung in"


In [96]:
billboard_all = pd.concat([pre_billboard, post_billboard], ignore_index=True)
print(billboard_all["era"].value_counts())

era
pre     130
post    130
Name: count, dtype: int64


In [100]:
billboard_all.rename(columns={"artists" :"artist_clean"}, inplace=True)

In [112]:
spotify_df['release_date'] = pd.to_datetime(
    spotify_df['release_date'],
    format='%m/%d/%y',
    errors='coerce'  # invalid parsing becomes NaT
)
spotify_df['release_year'] = spotify_df['release_date'].dt.year
billboard_all['year'] = pd.to_numeric(billboard_all['year'], errors='coerce')

In [115]:
merged_df = billboard_all.merge(
    spotify_df,
    left_on='song_clean',
    right_on='song_clean',
    how='inner',
    suffixes=('_billboard','_spotify')
)
merged_df

,Unnamed: 0,year_billboard,song,era,artist_clean_billboard,song_clean,id,name,artists,release_date,...,loudness,speechiness,tempo,valence,mode,key,explicit,duration_min,artist_clean_spotify,release_year
0,0,2000,"""Breathe""",pre,faith hill,breathe,06L1apH8kLF47dbhZ4Zg9A,BREATHE,['LEE HI'],2016-03-09,...,0.807313,-0.376546,0.222122,-0.613901,1,8,0,0.474654,lee hi,2016.0
1,0,2000,"""Breathe""",pre,faith hill,breathe,0uKrhHBDUQ4w5fv4qwOubP,Breathe,"['Télépopmusik', 'Angela McCluskey']",2001-01-01,...,0.730765,-0.348534,-0.096327,0.319767,1,1,0,0.418992,"télépopmusik,angela mccluskey",2001.0
2,0,2000,"""Breathe""",pre,faith hill,breathe,4Th9Lu8GK0mt5fZZ945FI7,Breathe,['Stick Figure'],2012-06-15,...,0.845697,-0.307850,-0.747026,0.879968,1,11,0,0.503437,stick figure,2012.0
3,0,2000,"""Breathe""",pre,faith hill,breathe,6bnVMdZObpNPUJ0vVFzkGn,Breathe,"['Jax Jones', 'Ina Wroldsen']",2020-06-12,...,0.861561,-0.341864,0.290727,0.460770,0,11,0,-0.145564,"jax jones,ina wroldsen",2020.0
4,0,2000,"""Breathe""",pre,faith hill,breathe,7fuoqve1SOEBmhc2PoQTF0,Breathe,['Fabolous'],2004-10-11,...,0.852149,1.253466,-1.103911,-0.084187,0,10,1,0.309531,fabolous,2004.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1215,253,2025,"""Lose Control""",post,teddy swims,lose control,0UaMYEvWZi0ZqiDOoHU3YI,Lose Control (feat. Ciara & Fat Man Scoop),"['Missy Elliott', 'Ciara', 'Fatman Scoop']",2005-07-04,...,0.828361,0.179686,0.277058,1.059080,0,4,1,-0.037447,"missy elliott,ciara,fatman scoop",2005.0
1216,253,2025,"""Lose Control""",post,teddy swims,lose control,1ulAlG3x3jWtCCK2SI4qyL,Lose Control (feat. Ciara & Fat Man Scoop),"['Missy Elliott', 'Ciara', 'Fatman Scoop']",2005-07-04,...,0.829395,0.086314,0.277090,0.963808,0,4,0,-0.028965,"missy elliott,ciara,fatman scoop",2005.0
1217,254,2025,"""Birds of a Feather""",post,billie eilish,birds of a feather,7AZY4swaZBthd2r0VP3qJe,Birds Of A Feather,['Gene Krupa'],2058-01-01,...,0.717986,-0.238488,1.597818,0.761831,0,5,0,-0.262393,gene krupa,2058.0
1218,255,2025,"""Beautiful Things""",post,benson boone,beautiful things,06wxyCQFJOT0bjvSPMQj7x,Beautiful Things,['Gungor'],NaT,...,0.810665,-0.458580,-1.136683,-1.410377,1,2,0,0.650219,gungor,NaN


In [122]:
merged_df['year_diff'] = (merged_df['release_year'] - merged_df['year_billboard']).abs()
merged_df = merged_df.sort_values('year_diff').drop_duplicates(subset=['song_clean'], keep='first')

In [123]:
print(len(billboard_all))
print(len(merged_df))

260
228


In [124]:
merged_df

,Unnamed: 0,year_billboard,song,era,artist_clean_billboard,song_clean,id,name,artists,release_date,...,speechiness,tempo,valence,mode,key,explicit,duration_min,artist_clean_spotify,release_year,year_diff
604,125,2012,"""Glad You Came""",pre,the wanted,glad you came,0dw7G0QDuW7UwyL0boHKMU,Glad You Came,['The Wanted'],2012-01-01,...,-0.165791,0.323532,-0.385247,0,7,0,-0.273818,the wanted,2012.0,0.0
1008,193,2019,"""Bad Guy""",post,billie eilish,bad guy,3yNZ5r3LKfdmjoS3gkhUCT,bad guy (with Justin Bieber),"['Billie Eilish', 'Justin Bieber']",2019-07-11,...,1.340168,0.589291,0.563664,1,0,0,-0.301405,"billie eilish,justin bieber",2019.0,0.0
1017,196,2019,"""7 Rings""",post,ariana grande,7 rings,6ocbgoVGwYJhOv1GgI9NsF,7 rings,['Ariana Grande'],2019-02-08,...,1.600276,0.751787,-0.781580,0,1,1,-0.435041,ariana grande,2019.0,0.0
619,128,2012,"""Starships""",pre,nicki minaj,starships,59RSrxkSKSixEWBAIvE5Wa,Starships,['Nicki Minaj'],2012-01-01,...,-0.127109,0.262315,0.834238,0,11,1,-0.171280,nicki minaj,2012.0,0.0
623,129,2012,"""What Makes You Beautiful""",pre,one direction,what makes you beautiful,4cluDES4hQEUhmXj6TXkSo,What Makes You Beautiful,['One Direction'],2012-05-25,...,-0.135779,0.261729,1.356330,1,4,0,-0.258981,one direction,2012.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1217,254,2025,"""Birds of a Feather""",post,billie eilish,birds of a feather,7AZY4swaZBthd2r0VP3qJe,Birds Of A Feather,['Gene Krupa'],2058-01-01,...,-0.238488,1.597818,0.761831,0,5,0,-0.262393,gene krupa,2058.0,33.0
64,10,2001,"""Hanging by a Moment""",pre,lifehouse,hanging by a moment,0wqOReZDnrefefEsrIGeR4,Hanging By A Moment,['Lifehouse'],NaT,...,-0.389218,0.247637,-0.370004,1,1,0,-0.126441,lifehouse,NaN,NaN
219,44,2004,"""The Way You Move""",pre,"outkast,sleepy brown",the way you move,5z7mYFfhw6N6f23VwrokJD,The Way You Move (feat. Sleepy Brown),"['OutKast', 'Sleepy Brown']",NaT,...,-0.317855,0.294567,0.392174,0,5,1,0.021372,"outkast,sleepy brown",NaN,NaN
567,118,2011,"""Moves like Jagger""",pre,"maroon 5,christina aguilera",moves like jagger,2LNbLA2JSQlv6NRz1xdZRj,Moves Like Jagger - Studio Recording From The ...,"['Maroon 5', 'Christina Aguilera']",NaT,...,-0.308517,0.360274,0.361687,1,9,0,-0.248546,"maroon 5,christina aguilera",NaN,NaN


In [131]:
def determine_era(year):
    if year < 2013:
        return "pre"
    else:
        return "post"

merged_df["era"] = merged_df["year_billboard"].apply(determine_era)
merged_df.to_csv("data/01-interim/merged-spotify-billboard", index=False)

display(top10_yearly.isna().sum())

year       0
song       0
era        0
artists    0
dtype: int64

In [132]:
print(merged_df["era"].unique())

['pre' 'post']


In [133]:
pre_streaming_features = merged_df[merged_df["era"] == 'pre']
post_streaming_features = merged_df[merged_df["era"] == 'post']

pre_streaming_features.shape

(129, 26)

In [134]:
post_streaming_features.shape

(99, 26)

### Dataset #3: Genius Lyrics Dataset

For this project, we sourced lyrics from the Genius Lyrics Database, filtering entries to match our selected sample of Billboard songs. The Genius database is a crowdsourced repository containing over 8 million lyrics, making it the largest publicly available resource for lyrical and musical knowledge. Contributions are often verified by the artists themselves, ensuring a high degree of accuracy not only for the lyrics but also for the annotations and analyses associated with each song. Using the LyricsGenius Python client, we retrieved lyrics corresponding to our sample of songs from both the pre- and post-streaming eras, enabling a consistent and structured dataset for analysis. These lyrics form the basis for our exploration of lyrical features, including sentiment, mood, and thematic content. In particular, we utilize a pretrained machine learning model provided by Nyckel, which classifies lyrics according to emotional tone and positivity/negativity, allowing us to quantitatively compare lyrical trends across different time periods.

Note: Running this section of the data wrangling requires access to a private API key for the Genius API. Please visit the link below. Put the link to the Github for all prompts then generate an API key. Replace the following lines with your API code. (REMINDER: do not push your personal API key to Github!!!)

https://genius.com/api-clients

In [18]:
import os
import pandas as pd
import re
os.environ["GENIUS_API_TOKEN"] = "tMjua3hBq2PhU2qwO5lvDwDC8FOG5QOA0fBoy4y1de7GKkVOm3WuvJ1GCy3C4eAd"

print("Genius API token set successfully!")

Genius API token set successfully!


In [19]:
#run pip install lyricsgenius first if you do not have it
import lyricsgenius

GENIUS_TOKEN = os.getenv("GENIUS_API_TOKEN")

if GENIUS_TOKEN is None:
    raise ValueError(
        "Genius API token not found. Please set GENIUS_API_TOKEN as an environment variable."
    )

genius = lyricsgenius.Genius(
    GENIUS_TOKEN,
    skip_non_songs=True,
    remove_section_headers=True,
    excluded_terms=["(Remix)", "(Live)"]
)

print("Genius client initialized successfully.")

Genius client initialized successfully.


In [119]:
def clean_artist(artist):
    return artist.split(",")[0].strip()
def clean_title(title):
    title = re.sub(r"\(.*?\)", "", title)  
    title = re.sub(r"- .*", "", title)     
    return title.strip()
    
pre_streaming_data = pd.read_csv("data/02-processed/pre_streaming_top10_annual")
post_streaming_data = pd.read_csv("data/02-processed/post_streaming_top10_annual")

pre_streaming_data

pre_streaming_data["genius_artist"] = pre_streaming_data["artists"].apply(clean_artist)
pre_streaming_data["genius_title"] = pre_streaming_data["song"].apply(clean_title)

post_streaming_data["genius_artist"] = post_streaming_data["artists"].apply(clean_artist)
post_streaming_data["genius_title"] = post_streaming_data["song"].apply(clean_title)

In [ ]:
lyrics_rows = []

# ---- PRE STREAMING ----
for idx, row in pre_streaming_data.iterrows():
    title = row["genius_title"]
    artist = row["genius_artist"]

    try:
        song = genius.search_song(title, artist, get_full_info=False)

        lyrics_rows.append({
            "title": row["Song"],
            "artist": row["artists_norm"],
            "lyrics": song.lyrics if song else None,
            "era": "pre"
        })

    except Exception as e:
        print(f"Error with {title}: {e}")
        lyrics_rows.append({
            "title": row["title"],
            "artist": row["artist"],
            "lyrics": None,
            "era": "pre"
        })


# ---- POST STREAMING ----
for idx, row in post_streaming_data.iterrows():
    title = row["genius_title"]
    artist = row["genius_artist"]

    try:
        song = genius.search_song(title, artist, get_full_info=False)

        lyrics_rows.append({
            "title": row["Song"],
            "artist": row["artists_norm"],
            "lyrics": song.lyrics if song else None,
            "era": "post"
        })

    except Exception as e:
        print(f"Error with {title}: {e}")
        lyrics_rows.append({
            "title": row["Song"],
            "artist": row["artist_norm"],
            "lyrics": None,
            "era": "post"
        })

# ---- CREATE FINAL DATAFRAME ----
final_df = pd.DataFrame(lyrics_rows)

## Results

### Exploratory Data Analysis

Instructions: replace the words in this subsection with whatever words you need to setup and preview the EDA you're going to do.   

Please explicitly load the fully wrangled data you will use from `data/02-processed`.  This is a good idea rather than forcing people to re-run the data getting / wrangling cells above.  Sometimes it takes a long time to get / wrangle data compared to reloading the fixed up dataset.

Carry out whatever EDA you need to for your project in the code cells below.  Because every project will be different we can't really give you much of a template at this point. But please make sure you describe the what and why in text here as well as providing interpretation of results and context.

Please note that you should consider the use of python modules in your work.  Any code which gets called repeatedly should be modularized. So if you run the same pre-processing, analysis or visualiazation on different subsets of the data, then you should turn that into a function or class.  Put that function or class in a .py file that lives in `modules/`.  Import the module you made and use it to get your work done.  For reference see `get_raw()` which is inside `modules/get_data.py`. 



#### Section 1 of EDA - Musical Feature Variability Across Pre-streaming and Post-streaming Eras

The results indicate that variability in nearly all musical features increased in the streaming era (2013–2026) relative to the pre-streaming era (2000–2012). The biggest change occurred in “instrumentalness”, which increased by over 59%, suggesting that there was quite a big expansion in the range of instrumental presence among top-charting songs. This may reflect increased production styles, and broader stylistic experimentation within songs that was facilitated by streaming platforms.
Other features such as acousticness (+15%), loudness (+10%), energy (+5.9%), valence (+5.7%), and tempo (+4%) also show pretty moderate increases in variability. These findings suggest that mainstream music in the streaming era actually has greater diversity in mood, production intensity, and rhythmic structure when compared to the pre-streaming era.
The only feature that decreased slightly in variability was explicitness (−1.79%), showing that while explicit content may have become more common overall, it has also become more standardized across popular songs.
In all, the evidence suggests that mainstream music has actually become more diverse in the streaming era rather than more homogenized. Streaming services, according to the data, seem to have led to an increase in the variation of popular music.


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

In [ ]:
features_00_13['era'] = 'Pre-streaming'
features_14_26['era'] = 'Post-streaming'

features_eda = pd.concat([features_00_13, features_14_26], ignore_index=True)
features_eda

In [ ]:
features_eda['chart_date'] = pd.to_datetime(features_eda['release_date_y'])
features_eda['year'] = features_eda['chart_date'].dt.year
features_eda['month'] = features_eda['chart_date'].dt.to_period('M')
features_eda = features_eda.drop(columns=['release_date_x','release_date_y','year_x','year_y'])

features_eda

In [ ]:
features_eda.groupby('month').size()

In [ ]:
first_chart_per_month = features_eda.groupby('month')['chart_date'].min().reset_index()
monthly_snapshot = features_eda.merge(first_chart_per_month, on=['month','chart_date'])

In [ ]:
monthly_snapshot.groupby('month').size().value_counts()

In [ ]:
complete_weeks = features_eda.groupby('chart_date').filter(lambda x: len(x) == 3)
print(complete_weeks.groupby('chart_date').size().value_counts())
complete_weeks.groupby('era')['chart_date'].nunique()

In [ ]:
complete_weeks.columns

Computing Weekly Variability

In [ ]:
weekly_explicit = complete_weeks.groupby(['chart_date','era'])['explicit'].mean().reset_index()
complete_weeks.groupby(['chart_date','era'])['key'].nunique()

In [ ]:
sns.boxplot(data=weekly_explicit, x='era', y='explicit')
plt.title("Weekly Explicitness in Top-3 Songs by Era")
plt.show()

In [ ]:
weekly_key_div = complete_weeks.groupby(['chart_date','era'])['key'].nunique().reset_index()
weekly_key_div.rename(columns={'key':'num_unique_keys'}, inplace=True)

In [ ]:
sns.boxplot(data=weekly_key_div, x='era', y='num_unique_keys')
plt.title("Weekly Key Diversity in Top-3 Songs by Era")
plt.show()

In [ ]:
features = ['acousticness', 'danceability', 'energy', 'loudness', 'tempo', 'valence', 'duration_min']
weekly_var = (
    complete_weeks
    .groupby(['chart_date','era'])[features]
    .std()
    .reset_index()
)

weekly_long = weekly_var.melt(
    id_vars=['chart_date','era'],
    var_name='feature',
    value_name='variability'
)

In [ ]:
plt.figure(figsize=(14,6))
sns.boxplot(data=weekly_long, x='feature', y='variability', hue='era')
plt.xticks(rotation=45)
plt.title("Weekly Top-3 Variability by Feature and Era")
plt.tight_layout()
plt.show()

In [ ]:
pre = weekly_var[weekly_var['era']=='Pre-streaming']
post = weekly_var[weekly_var['era']=='Post-streaming']

In [ ]:
results = []

# Bonferroni correction
alpha = 0.05
corrected_alpha = alpha / len(features)

def cohens_d(x, y):
    pooled_sd = np.sqrt((x.var() + y.var()) / 2)
    return (x.mean() - y.mean()) / pooled_sd

for feature in features:
    # Shapiro-Wilk test for normality
    p_pre = stats.shapiro(pre[feature])[1]
    p_post = stats.shapiro(post[feature])[1]
    
    normal = (p_pre > 0.05) and (p_post > 0.05)
    
    if normal:
        test_name = "Welch t-test"
        stat, p_val = stats.ttest_ind(pre[feature], post[feature], equal_var=False)
    else:
        test_name = "Mann-Whitney U"
        stat, p_val = stats.mannwhitneyu(pre[feature], post[feature])
    
    d = cohens_d(pre[feature], post[feature])
    
    # Mean SD per era
    mean_pre = pre[feature].mean()
    mean_post = post[feature].mean()
    
    results.append({
        "Feature": feature,
        "Pre-streaming mean SD": mean_pre,
        "Post-streaming mean SD": mean_post,
        "Test": test_name,
        "Stat": stat,
        "p-value": p_val,
        "Bonferroni sig.": p_val < corrected_alpha,
        "Cohen's d": d
    })

results_df = pd.DataFrame(results)
pd.set_option('display.float_format', lambda x: '%.3f' % x)
print(results_df)

#### Section 2 of EDA if you need it  - please give it a better title than this

Some more words and stuff.  Remember notebooks work best if you interleave the code that generates a result with properly annotate figures and text that puts these results into context.

In [ ]:
## YOUR CODE HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION

## Ethics

Here is a [list of real world examples](https://deon.drivendata.org/examples/) for each item in the checklist that can refer to.

[![Deon badge](https://img.shields.io/badge/ethics%20checklist-deon-brightgreen.svg?style=popout-square)](http://deon.drivendata.org/)

### A. Data Collection
 - [ ] **A.1 Informed consent**: If there are human subjects, have they given informed consent, where subjects affirmatively opt-in and have a clear understanding of the data uses to which they consent?
 - [X] **A.2 Collection bias**: Have we considered sources of bias that could be introduced during data collection and survey design and taken steps to mitigate those?
> Because we are only looking at the top 3 Billboard Hot 100 songs, selection bias may occur due to factors such as audience preference, popularity, and marketing power. Selection bias will lean towards successful artists and popularized genres that hold these factors, making the scope of songs analyzed less diverse because we will be excluding various genres that are less popular and more niche.
 - [X] **A.3 Limit PII exposure**: Have we considered ways to minimize exposure of personally identifiable information (PII) for example through anonymization or not collecting information that isn't relevant for analysis?
> The data that will be analyzed in this project comes from publicly available Billboard Hot 100 chart information. The analysis focuses only on the top three songs per period and does not collect or use any personal or private data. As a result, there is no exposure of personally identifiable information.
 - [ ] **A.4 Downstream bias mitigation**: Have we considered ways to enable testing downstream results for biased outcomes (e.g., collecting data on protected group status like race or gender)?

### B. Data Storage
 - [ ] **B.1 Data security**: Do we have a plan to protect and secure data (e.g., encryption at rest and in transit, access controls on internal users and third parties, access logs, and up-to-date software)?
 - [ ] **B.2 Right to be forgotten**: Do we have a mechanism through which an individual can request their personal information be removed?
 - [ ] **B.3 Data retention plan**: Is there a schedule or plan to delete the data after it is no longer needed?

### C. Analysis
 - [X] **C.1 Missing perspectives**: Have we sought to address blindspots in the analysis through engagement with relevant stakeholders (e.g., checking assumptions and discussing implications with affected communities and subject matter experts)?
> Since we will be analyzing the Billboard Hot 100 chart and more specifically the top 3 songs of the chart, it is likely that the top 3 songs will have alike genres, audiences, and artists. This results in the analysis overlooking qualitative factors that play a role in music like culture, sociopolitical influences, and artistic expression that quantitative factors fail to capture. This fails to include the analysis of songs and genres who do not qualify as mainstream charting music.
 - [X] **C.2 Dataset bias**: Have we examined the data for possible sources of bias and taken steps to mitigate or address these biases (e.g., stereotype perpetuation, confirmation bias, imbalanced classes, or omitted confounding variables)?
> The Billboard Hot 100 chart is composed of songs that are most popular in the U.S. that  lead to geographic and cultural biases in the dataset. This overrepresentation of U.S. trends in music affects the conclusion being made about musical homogenization because it excludes global trends, artists, and listening patterns in music that is not part of mainstream U.S. music
 - [ ] **C.3 Honest representation**: Are our visualizations, summary statistics, and reports designed to honestly represent the underlying data?
 - [X] **C.4 Privacy in analysis**: Have we ensured that data with PII are not used or displayed unless necessary for the analysis?
> The analysis conducted to form a response to the research question is based solely on songs and music trends that are publicly shared and are not focused on private individuals.
 - [X] **C.5 Auditability**: Is the process of generating the analysis well documented and reproducible if we discover issues in the future?
> The analysis includes data sources that are traceable and can be accessed by anyone who later on wishes to analyze it too. The well documented data also includes preprocessing steps and analytical methods that are used to reach a conclusion. Together this project can be reanalyzed by later on researchers, allowing them to either replicate results or identify potential errors.

### D. Modeling
 - [ ] **D.1 Proxy discrimination**: Have we ensured that the model does not rely on variables or proxies for variables that are unfairly discriminatory?
 - [ ] **D.2 Fairness across groups**: Have we tested model results for fairness with respect to different affected groups (e.g., tested for disparate error rates)?
 - [X] **D.3 Metric selection**: Have we considered the effects of optimizing for our defined metrics and considered additional metrics?
> The variables that will be used to measure similarities over the years are mainly audio based such as tempo, danceability, and instrumentals. We do not include other factors such as emotion and cultural significance which can also affect our conclusion.  
 - [ ] **D.4 Explainability**: Can we explain in understandable terms a decision the model made in cases where a justification is needed?
 - [X] **D.5 Communicate limitations**: Have we communicated the shortcomings, limitations, and biases of the model to relevant stakeholders in ways that can be generally understood?
> As previously stated, the data sets are limited to U.S. music trends, mainstream songs, quantitative features and exclude niche genres, cultural and emotional significance, and qualitative features.

### E. Deployment
 - [ ] **E.1 Monitoring and evaluation**: Do we have a clear plan to monitor the model and its impacts after it is deployed (e.g., performance monitoring, regular audit of sample predictions, human review of high-stakes decisions, reviewing downstream impacts of errors or low-confidence decisions, testing for concept drift)?
 - [X] **E.2 Redress**: Have we discussed with our organization a plan for response if users are harmed by the results (e.g., how does the data science team evaluate these cases and update analysis and models to prevent future harm)?
> Since the project only focuses on the composition of songs and analyzes factors that observe how songs are homogenized throughout the years,no users will be directly harmed with the results of the model, making redress in inapplicable. 
 - [ ] **E.3 Roll back**: Is there a way to turn off or roll back the model in production if necessary?
 - [X] **E.4 Unintended use**: Have we taken steps to identify and prevent unintended uses and abuse of the model and do we have a plan to monitor these once the model is deployed?
> Looking to see if songs have become more homogenized over the years, there is the possibility that people may use the results of the model to criticize mainstream songs and popular genres of music. We acknowledge some may use the results of the model to target artists or nowadays popular music and will work carefully and respectfully to form a carefully presented analysis and conclusion, focusing on limitations and avoiding judgement to artists and their music.

## Team Expectations 

* Communicate with rest of the team members within at most twelve hours of initial communication.
* Be open to everyone's ideas and inputs and make decisions based on maximum benefit to the project.
* Members with more experience with Git, data analysis and/or data visualization will assist members with less experience, so everyone is on the same page and understands how to contribute to the project. 
* Tasks are divided evenly so that no one is taking on too much work, even if that means some members specialize in some tasks more than others.
* Members will meet once either in-person or online before important deadlines/checkpoints to make sure that project is on track.
* Should a conflict of interest arise, members will resolve it respectfully and not deter the progress of the project.

## Project Timeline Proposal

 | Meeting Date  | Meeting Time| Completed Before Meeting  | Discuss at Meeting |
|---|---|---|---|
| 2/11  |  3:30 PM | Planning Project Proposal Sections  | All members will communicate with each other to complete their sections of the Project Proposal. | 
| 2/18  |  3:30 PM |  Edit, finalize, and submit proposal; search for datasets | Discuss wrangling and assign parts for the data visualization. Complete data checkpoint. | 
| 2/25  | 3:30 PM  | Import and Wrangle Data   | Discuss wrangling and analytical approaches; Assign group members to lead each specific part |
| 3/4  | 3:30 PM  | Edit and Finalize Analysis | Members will finalize the analysis and assign parts for results/conclusion/discussion. Two members will start working on the video component.    |
| 3/11  | 3 PM  | Finalize all components of the Project, including the video portion. | Discuss/edit full project |
| 3/18  | 3 PM  | Proofread/Overlook Project| Turn in Project by 11:59. |